homeassistant/components/ecowitt/sensor.py

```diff
@@ -252,22 +266,79 @@ def _new_sensor(sensor: EcoWittSensor) -> None:
             "hourlyrainmm",
             "hourlyrainin",
         ):
-            description = dataclasses.replace(
+            description = replace(
                 description,
                 state_class=SensorStateClass.MEASUREMENT,
             )
 
-        async_add_entities([EcowittSensorEntity(sensor, description)])
+        return EcowittSensorEntity(sensor, description)
+
+    def _new_sensor(*sensors: EcoWittSensor) -> None:
+        """Add new sensor entities if they don't already exist."""
+        new_sensor_entities: list[EcowittSensorEntity] = []
+        for sensor in sensors:
+            if sensor.key in added_keys:
+                _LOGGER.debug("Sensor already added: %s", sensor.key)
+                continue
+
+            if sensor.stype not in ECOWITT_SENSORS_MAPPING:
+                continue
+
+            # Ignore metrics that are not supported by the user's locale
+            if sensor.stype in _METRIC and hass.config.units is not METRIC_SYSTEM:
+                continue
+            if (
+                sensor.stype in _IMPERIAL
+                and hass.config.units is not US_CUSTOMARY_SYSTEM
+            ):
+                continue
 
+            mapping = ECOWITT_SENSORS_MAPPING[sensor.stype]
+            entity = _generate_entity(sensor, mapping)
+            new_sensor_entities.append(entity)
+            added_keys.add(sensor.key)
+        async_add_entities(new_sensor_entities)
+
+    entity_registry = er.async_get(hass)
+    entries = er.async_entries_for_config_entry(entity_registry, entry.entry_id)
+
+    # Restore entities from the entity registry
+    for config_entry in entries:
+        if config_entry.original_name is not None:
+            config_entry_unique_id = config_entry.unique_id.split("-")[0]
+            config_entry_key = config_entry.unique_id.split("-")[1]
+            sensor_info = SENSOR_INFO_BY_NAME.get(config_entry.original_name)
+            if sensor_info is None:
+                return
+            sensor_key_name = sensor_info.key
+            station = EcoWittStation(
+                **entry.data["station"],
+            )
+            sensor = ecowitt.sensors[f"{config_entry_unique_id}.{sensor_key_name}"] = (
+                EcoWittSensor(
+                    name=config_entry.original_name,
+                    key=config_entry_key,
+                    stype=sensor_key_name,  # type: ignore[arg-type]
+                    station=station,
+                )
+            )
```

Comment:
This should be done in the library to reduce errors, such as when you forget to add the sensor to the EcoWittListener.
In your PR the sensor will be created EcoWittSensor.
Creating them here is hacky